x/oracle/keeper/abci.go

```diff
@@ -6,164 +6,197 @@ import (
 
-		// iterate over the data votes
-		k.IterateOracleDataVotes(ctx, func(val sdk.AccAddress, msg *types.MsgOracleDataVote) bool {
-			// save a voted array
-			voted = append(voted, val.String())
+		// all aggregated oracle data for the current voting period
+		aggregateMap = make(map[string]types.OracleData)
+	)
 
-			// remove the miss counter for validators who have voted
-			k.DeleteMissCounter(ctx, val)
+	// iterate over the data votes
+	// TODO: only iterate on the last voting period
+	k.IterateOracleDataVotes(ctx, func(validatorAddr sdk.AccAddress, vote types.OracleVote) bool {
+		// mark the validator voting as true
+		validatorVotesMap[validatorAddr.String()] = true
 
-			// find total voting votedPower
-			votedPower += k.stakingKeeper.Validator(ctx, sdk.ValAddress(val)).GetConsensusPower()
+		// remove the miss counter in the current voting window for validators who have already voted
+		k.DeleteMissCounter(ctx, validatorAddr)
 
-			// save the oracle data for later processing
-			for _, oda := range msg.OracleData {
-				od, err := types.UnpackOracleData(oda)
-				if err != nil {
-					panic(err)
-				}
+		// find total voting votedPower
+		validator := k.stakingKeeper.Validator(ctx, sdk.ValAddress(validatorAddr))
+		if validator == nil {
+			// validator nor found, continue with next vote
+			return false
+		}
 
-				detailedMap[od.Type()][val.String()] = od
-				collectionMap[od.Type()] = append(collectionMap[od.Type()], od)
-			}
+		votedPower += validator.GetConsensusPower()
 
-			// delete the vote as we no longer require it
-			k.DeleteOracleDataVote(ctx, val)
+		// Safety check. Already validated on msg validation
+		if vote.Feed == nil || len(vote.Feed.OracleData) == 0 {
+			k.Logger(ctx).Debug("attempted to process empty oracle data from feed", "validator", validatorAddr.String())
 			return false
-		})
-
-		// iterate over the full list of validators to increment miss counters
-		totalPower := int64(0)
-		for _, val := range k.stakingKeeper.GetBondedValidatorsByPower(ctx) {
-			totalPower += val.GetConsensusPower()
-			valaddr := sdk.AccAddress(val.GetOperator())
-			if !contains(voted, valaddr.String()) {
-				k.IncrementMissCounter(ctx, valaddr)
+		}
+
+		// save the oracle data for later processing
+		for _, oracleDataAny := range vote.Feed.OracleData {
+			oracleData, err := types.UnpackOracleData(oracleDataAny)
+			if err != nil {
+				// NOTE: this should never panic as the oracle data had already been checked before
+				// setting it to store.
+				panic(err)
 			}
+
+			// add oracle data to maps
+			submittedFeedMap[oracleData.Type()][validatorAddr.String()] = oracleData
+
 		}
 
-		// if the voted_power/total_power > params.VoteThreshold then we store the averages in the store
-		storeAverages := sdk.NewDec(votedPower).Quo(sdk.NewDec(totalPower)).GT(params.VoteThreshold)
+		// delete the vote as it has already been processed
+		// TODO: consider keeping the votes for a few voting windows in order to
```

Comment:
I think saving the votes will be too expensive. cc @zmanian

Reply:
agree, maybe for 20 blocks should be sufficient?